source/Octopus.Tentacle/Services/Scripts/ScriptServiceV2.cs

```diff
@@ -0,0 +1,134 @@
+﻿using System.Collections.Concurrent;
+using System.Threading;
+using Octopus.Diagnostics;
+using Octopus.Tentacle.Contracts;
+using Octopus.Tentacle.Scripts;
+
+namespace Octopus.Tentacle.Services.Scripts
+{
+    [Service]
+    public class ScriptServiceV2 : IScriptServiceV2
+    {
+        readonly IShell shell;
+        readonly IScriptWorkspaceFactory workspaceFactory;
+        readonly IScriptStateWorkspaceFactory stateWorkspaceFactory;
+        readonly ISystemLog log;
+        readonly ConcurrentDictionary<ScriptTicket, RunningScriptAndCancellationToken> scripts = new();
+
+        public ScriptServiceV2(
+            IShell shell,
+            IScriptWorkspaceFactory workspaceFactory,
+            IScriptStateWorkspaceFactory stateWorkspaceFactory,
+            ISystemLog log)
+        {
+            this.shell = shell;
+            this.workspaceFactory = workspaceFactory;
+            this.stateWorkspaceFactory = stateWorkspaceFactory;
+            this.log = log;
+        }
+
+        public ScriptTicket StartScript(StartScriptCommand command, ScriptTicket ticket)
+        {
+            var workspace = workspaceFactory.PrepareWorkspace(command, ticket);
```

Comment:
Should this method ensure that it is only called once? Prepare workspace looks like something that on a second call would override the files it already wrote down.

Perhaps it would make sense for this class to have some sort of lock[s] controlling concurrency to the same ScriptTicket.

If this is called again with the same ScriptTicket while this method is being run with that same script ticket the second should just take the result of the first. :D